Alan/pylint rcfile

alanngai opened the pull request:
updated pr with last set of feedback from https://github.com/twitter/heron/pull/780.  Remaining potential issues

heron/ui/src/python/handler/api/topology.py: 
W:231, 0: FIXME: is there a purpose to saving the return value of this? (fixme)
W:234, 0: FIXME: estate not defined (fixme)
E:240,13: Undefined variable 'estate' (undefined-variable)
W:232, 4: Unused variable 'scheduler_location' (unused-variable)

---- pylint output for heron/cli ------------------------------------------------------------------
************* Module main
I: 42, 0: Locally disabling protected-access (W0212) (locally-disabled)
I: 64, 0: Locally disabling bad-super-call (E1003) (locally-disabled)
************* Module color
I: 18, 0: Locally disabling invalid-name (C0103) (locally-disabled)
************* Module log
I: 19, 0: Locally disabling invalid-name (C0103) (locally-disabled)
************* Module activate
I: 24, 0: Locally disabling unused-argument (W0613) (locally-disabled)
************* Module args
I: 24, 0: Locally disabling protected-access (W0212) (locally-disabled)
************* Module deactivate
I: 24, 0: Locally disabling unused-argument (W0613) (locally-disabled)
************* Module execute
W: 87, 0: TODO: We need to extract heron-core to separate folder. (fixme)
************* Module help
I: 27, 0: Locally disabling protected-access (W0212) (locally-disabled)
I: 40, 0: Locally disabling unused-argument (W0613) (locally-disabled)
************* Module jars
W: 21, 0: TODO, make the jars independent version free (fixme)
W: 33, 0: TODO, make the jars independent version free (fixme)
************* Module kill
I: 24, 0: Locally disabling unused-argument (W0613) (locally-disabled)
************* Module opts
I: 19, 0: Locally disabling invalid-name (C0103) (locally-disabled)
I: 19, 0: Locally disabling global-variable-not-assigned (W0602) (locally-disabled)
I: 19, 0: Locally disabling global-statement (W0603) (locally-disabled)
************* Module restart
I: 50, 0: Locally disabling unused-argument (W0613) (locally-disabled)
************* Module submit
I:234, 0: Locally disabling unused-argument (W0613) (locally-disabled)
************* Module utils
I: 86, 0: Locally disabling protected-access (W0212) (locally-disabled)
R:218, 0: Too many branches (14/12) (too-many-branches)
************* Module version
I: 34, 0: Locally disabling unused-argument (W0613) (locally-disabled)


Report
======
618 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |16     |16         |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |2      |2          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |2      |2          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|function |73     |73         |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+



External dependencies
---------------------
::

    utils (help)



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |732    |42.43 |732      |=          |
+----------+-------+------+---------+-----------+
|docstring |442    |25.62 |442      |=          |
+----------+-------+------+---------+-----------+
|comment   |339    |19.65 |339      |=          |
+----------+-------+------+---------+-----------+
|empty     |212    |12.29 |212      |=          |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |0      |0        |=          |
+-----------+-------+---------+-----------+
|refactor   |1      |1        |=          |
+-----------+-------+---------+-----------+
|warning    |3      |3        |=          |
+-----------+-------+---------+-----------+
|error      |0      |0        |=          |
+-----------+-------+---------+-----------+



% errors / warnings by module
-----------------------------

+--------+------+--------+---------+-----------+
|module  |error |warning |refactor |convention |
+========+======+========+=========+===========+
|jars    |0.00  |66.67   |0.00     |0.00       |
+--------+------+--------+---------+-----------+
|execute |0.00  |33.33   |0.00     |0.00       |
+--------+------+--------+---------+-----------+
|utils   |0.00  |0.00    |100.00   |0.00       |
+--------+------+--------+---------+-----------+



Messages
--------

+------------------+------------+
|message id        |occurrences |
+==================+============+
|locally-disabled  |17          |
+------------------+------------+
|fixme             |3           |
+------------------+------------+
|too-many-branches |1           |
+------------------+------------+



Global evaluation
-----------------
Your code has been rated at 9.94/10 (previous run: 9.94/10, +0.00)

---- pylint output for heron/ui ------------------------------------------------------------------
************* Module main
I: 39, 0: Locally disabling line-too-long (C0301) (locally-disabled)
I: 39, 0: Locally disabling bad-whitespace (C0326) (locally-disabled)
I: 93, 0: Locally disabling unused-argument (W0613) (locally-disabled)
W: 37, 0: TODO: Change these to query string parameters, since (fixme)
************* Module args
I: 19, 0: Locally disabling protected-access (W0212) (locally-disabled)
I: 42, 0: Locally disabling bad-super-call (E1003) (locally-disabled)
************* Module heron
I: 46, 0: Locally disabling invalid-name (C0103) (locally-disabled)
I:617, 0: Locally disabling no-self-use (R0201) (locally-disabled)
************* Module metrics
I: 22, 0: Locally disabling invalid-name (C0103) (locally-disabled)
************* Module topology
W:231, 0: FIXME: is there a purpose to saving the return value of this? (fixme)
W:234, 0: FIXME: estate not defined (fixme)
E:240,13: Undefined variable 'estate' (undefined-variable)
W:232, 4: Unused variable 'scheduler_location' (unused-variable)
************* Module graph
I: 26, 0: Locally disabling invalid-name (C0103) (locally-disabled)
************* Module utils
I: 15, 0: Locally disabling invalid-name (C0103) (locally-disabled)
************* Module notfound
I: 19, 0: Locally disabling unused-argument (W0613) (locally-disabled)
I:239, 0: Locally disabling attribute-defined-outside-init (W0201) (locally-disabled)
I:273, 0: Locally disabling attribute-defined-outside-init (W0201) (locally-disabled)
W: 71, 0: TODO: get the list of clusters from tracker and fetch the topologies (fixme)
************* Module log
I: 19, 0: Locally disabling invalid-name (C0103) (locally-disabled)
R:  1, 0: Similar lines in 2 files
==heron:531
==topology:153
      cluster=cluster,
      environ=environ,
      topology=topology,
      container=container,
      path=path
    ) (duplicate-code)
R:  1, 0: Similar lines in 2 files
==heron:505
==topology:210
      cluster=cluster,
      environ=environ,
      topology=topology,
      container=container,
      path=path, (duplicate-code)


Report
======
614 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |21     |21         |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |31     |31         |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |39     |39         |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|function |33     |33         |=          |100.00      |3.03     |
+---------+-------+-----------+-----------+------------+---------+



External dependencies
---------------------
::

    base (topology)
    fetch (heron)



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |939    |44.74 |939      |=          |
+----------+-------+------+---------+-----------+
|docstring |533    |25.39 |533      |=          |
+----------+-------+------+---------+-----------+
|comment   |341    |16.25 |341      |=          |
+----------+-------+------+---------+-----------+
|empty     |286    |13.63 |286      |=          |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |11    |11       |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.529 |0.529    |=          |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |0      |0        |=          |
+-----------+-------+---------+-----------+
|refactor   |2      |2        |=          |
+-----------+-------+---------+-----------+
|warning    |5      |5        |=          |
+-----------+-------+---------+-----------+
|error      |1      |1        |=          |
+-----------+-------+---------+-----------+



% errors / warnings by module
-----------------------------

+---------+------+--------+---------+-----------+
|module   |error |warning |refactor |convention |
+=========+======+========+=========+===========+
|topology |0.00  |20.00   |0.00     |0.00       |
+---------+------+--------+---------+-----------+
|main     |0.00  |20.00   |0.00     |0.00       |
+---------+------+--------+---------+-----------+
|log      |0.00  |0.00    |100.00   |0.00       |
+---------+------+--------+---------+-----------+



Messages
--------

+-------------------+------------+
|message id         |occurrences |
+===================+============+
|locally-disabled   |14          |
+-------------------+------------+
|fixme              |4           |
+-------------------+------------+
|duplicate-code     |2           |
+-------------------+------------+
|unused-variable    |1           |
+-------------------+------------+
|undefined-variable |1           |
+-------------------+------------+



Global evaluation
-----------------
Your code has been rated at 9.80/10 (previous run: 9.80/10, +0.00)

saileshmittal commented:
@alanngai , As I mentioned before, we are not following any particular python code styles, but we will be following [twitter guidelines](https://github.com/twitter/commons/blob/master/src/python/twitter/common/styleguide.md) eventually. One of the things I wanted to mention is the 2-space indentation, that you have already done, and 4-space indent for wrapped lines, like lines following an open bracket. Can you add that check as well?

saileshmittal commented on tools/python/checkstyle.ini:
```diff
@@ -0,0 +1,379 @@
+indent-string='  '
+
+# Number of spaces of indent required inside a hanging  or continued line.
+indent-after-paren=2
```
4 spaces.

alanngai replied:
fixed with latest commit

saileshmittal commented:
Otherwise LGTM. Can you also fix why travis build is failing?

saileshmittal commented on heron/cli/src/python/BUILD:
```diff
@@ -17,7 +17,8 @@ pex_library(
         "utils.py",
         "version.py",
-    ], 
+        "cli_helper.py",
```
Put it in alphabetical order.

In other places, we used globs so that we don't need to add every file manually.

```
srcs = glob(
        ["**/*.py"],
        exclude = ["main.py"],
    ),
```

I won't mind doing that here either.

alanngai commented:
The following are still issues.  I'll go ahead and open up a separate ticket

heron/ui/src/python/handler/api/topology.py: 
W:231, 0: FIXME: is there a purpose to saving the return value of this? (fixme)
W:234, 0: FIXME: estate not defined (fixme)
E:240,13: Undefined variable 'estate' (undefined-variable)
W:232, 4: Unused variable 'scheduler_location' (unused-variable)

alanngai commented:
build is still busted.  The build log indicates a problem executing a test topology, but doesn't say exactly what failed or why (see below).  The only thing I can think of is the addition of checkstyle.ini inside of tools/python that I added, and maybe that needs to be packaged in somehow.  I'm not familiar with how the build and test system works to do this.  Will need some help on this.  For reference, part of this PR was to modify the tools/java/.../PythonCheckstyle.java:84 where I added an additional command line parameter to read the checkstyle.ini file

      // Create and run the command
      List<String> commandBuilder = new ArrayList<>();
      commandBuilder.add(pylintFile);
      commandBuilder.add("--rcfile=" + PYLINT_RCFILE); // tools/python/checkstyle.ini
      commandBuilder.addAll(sourceFiles);
      runLinter(commandBuilder);

this seems to work fine locally when I run the linter, but maybe it breaks the build somehow?

## build output #######################################################
Heron is now installed!
Make sure you have "/home/travis/bin" in your path. 
See http://heronstreaming.io/docs/getting-started.html for how to use Heron.
heron.build.version : 0.14.0-1083-g85bfa2b
heron.build.time : Thu Jun  2 22:41:52 UTC 2016
heron.build.timestamp : 1464907312000
heron.build.host : testing-gce-e46a86b8-e324-4d1f-b559-f2aca088a128
heron.build.user : travis
heron.build.git.revision : 85bfa2bf893768f7e508ab45c2fd1188bd76f826
heron.build.git.status : Modified
+python integration-test/src/python/local_test_runner/main.py
INFO:root:Submitting topology
INFO:root:Submitting topology: 
INFO:root:[u'/home/travis/bin/heron', 'submit', '--verbose', '--', u'local', u'/home/travis/build/twitter/heron/bazel-genfiles/integration-test/src/java/local-integration-tests.jar', u'com.twitter.heron.local_integration_test.topology.local_readwrite.LocalReadWriteTopology', u'IntegrationTest_LocalReadWriteTopology', u'/home/travis/.herondata/topologies/local/travis/IntegrationTest_LocalReadWriteTopology/testing.txt', u'/home/travis/.herondata/topologies/local/travis/IntegrationTest_LocalReadWriteTopology/testing2.txt', '12']
{'config_property': [], 'topology-file-name': '/home/travis/build/twitter/heron/bazel-genfiles/integration-test/src/java/local-integration-tests.jar', 'override_config_file': '/tmp/tmpnKCfnQ/override.yaml', 'config_path': '/home/travis/.heron/conf/local', 'subcommand': 'submit', 'deploy_deactivated': False, 'environ': 'default', 'cluster': 'local', 'topology_main_jvm_property': [], 'role': 'travis', 'verbose': 'True', 'topology-class-name': 'com.twitter.heron.local_integration_test.topology.local_readwrite.LocalReadWriteTopology'}
ERROR: Unable to execute topology main class
INFO: Elapsed time: 0.001s.
INFO:root:Submitted topology
INFO:root:Successfully submitted KILL_TMASTER topology
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated

kramasamy commented:
@nlu90 - can you check and give some pointers to why the tests have failed?

nlu90 commented:
@alanngai There's not much useful information in the travis ci log.

Can you commit all your changes and push it to a branch on your fork so that I can manually check what's going on?

nlu90 commented:
Hi @alanngai, based on the branch of  `alan/pylint-rcfile` from your heron fork. I found the following problems:

1. In BUILD file: `"help.py",` should be `"cli_helper.py"` on line 12

2. In `main.py`: `import heron.cli.src.python.help as cli_help` should be `import heron.cli.src.python.cli_helper as cli_helper` on line 26, and change the following occurrence of `cli_help` to `cli_helper`.

3. In `main.py`: `cli_help.create_parser(subparsers)` only provides one argument while the method in `cli_helper.py` requires 3 parameters: `def create_parser(subparsers, action, help):`. I think we should either remove this line, or find the proper arguments for it. 

4. In `submit.py`: `javaDefines=cl_args['topology_main_jvm_property'])` on line 175 should be `java_defines=cl_args['topology_main_jvm_property'])`

5. In `submit.py`: `javaDefines=[]` on line 106 should be `java_defines=[]`

After that, I can launch launch a topology successfully. But there might be other problems.

nlu90 commented:
I tried with other heron operations and  found following problems: 

`kill` is not invoked correctly.
```bash
[tw-mbp-nlu heron (alan/pylint-rcfile)]$ heron kill local nlu-ex
INFO: heron_class 10
INFO: heron_class 20
[2016-06-09 14:23:44 -0700] com.twitter.heron.spi.utils.TMasterUtils SEVERE:  Topology is already activateed  
INFO: Successfully activate topology 'nlu-ex'
INFO: Elapsed time: 2.971s.
```

`deactivate` is not invoked correctly.
```bash
[tw-mbp-nlu heron (alan/pylint-rcfile)]$ heron deactivate local nlu-ex
INFO: heron_class 10
INFO: heron_class 20
[2016-06-09 14:27:26 -0700] com.twitter.heron.spi.utils.TMasterUtils SEVERE:  Topology is already activateed  
INFO: Successfully activate topology 'nlu-ex'
INFO: Elapsed time: 2.935s.
```

They are all bind to `activate` instead.

Could you fix those problems and let me know?

nlu90 commented:
@alanngai any updates?

alanngai commented:
apologies, will work on it later this week and/or weekend.  my plate is rather full this week

> On Jun 15, 2016, at 4:11 PM, Neng Lu <notifications@github.com> wrote:
> 
> @alanngai <https://github.com/alanngai> any updates?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/twitter/heron/pull/849#issuecomment-226346655>, or mute the thread <https://github.com/notifications/unsubscribe/ABk-epnzu0HKGi4OWSEtmlpzXZ7jzhQ_ks5qMIavgaJpZM4IsEM3>.
>

alanngai commented:
@nlu90, @kramasamy looks like everything passed finally.  ready to be merged

nlu90 commented:
👍

nlu90 commented on heron/cli/src/python/deactivate.py:
```diff
@@ -11,66 +11,25 @@
+  :param unknown_args:
+  :return:
+  '''
+  return cli_helper.run.run_helper(command, parser, cl_args, unknown_args, "deactivate topology")
```
`cli_helper.run.run_helper` should be `cli_helper.run`

nlu90 commented on heron/cli/src/python/deactivate.py:
```diff
@@ -32,4 +32,4 @@ def run(command, parser, cl_args, unknown_args):
   :return:
   '''
-  return cli_helper.run.run_helper(command, parser, cl_args, unknown_args, "deactivate topology")
+  return cli_helper.run_helper(command, parser, cl_args, unknown_args, "deactivate topology")
```
should be `cli_helper.run`

alanngai replied:
@nlu90: fixed and merge latest from master

nlu90 replied:
@alanngai I'm looking at the commit 7eb8cdf. Seems the typo is not fixed yet

alanngai replied:
oops, i was looking at the redundant run.run_helper.  Apologies

nlu90 commented on heron/ui/src/python/handlers/common/__init__.py:
```diff
@@ -1,4 +1,5 @@
 
 from heron.ui.src.python.handlers.common.consts import *
-from heron.ui.src.python.handlers.common.utils  import *
+from heron.ui.src.python.handlers.common.utils import *
```
On line 151 of file handlers/api/topology.py, `diameter = common.graph.TopologyDAG(lplan).diameter()`, `graph` is also used as member of the common module. We need to add one line here to include it in the common module `from heron.ui.src.python.handlers.common.graph import *`.

nlu90 commented:
The heron-cli part looks fine. But for the heron-ui/heron-tracker part, I see the following errors when looking around the topology page. @alanngai Can you fix the issue?
```
21 Jun 2016 11:44:33-ERROR:web.py:1407: Uncaught exception GET /topologies/metrics/timeline?cluster=local&environ=default&topology=nlu-ex&metric=gc&component=*&instance=*&starttime=1466523840&endtime=1466534640 (::1)
HTTPServerRequest(protocol='http', host='localhost:8889', method='GET', uri='/topologies/metrics/timeline?cluster=local&environ=default&topology=nlu-ex&metric=gc&component=*&instance=*&starttime=1466523840&endtime=1466534640', version='HTTP/1.1', remote_ip='::1', headers={'Accept-Language': 'en-US,en;q=0.8,zh-CN;q=0.6,zh;q=0.4,ja;q=0.2,zh-TW;q=0.2', 'Accept-Encoding': 'gzip, deflate, sdch', 'Host': 'localhost:8889', 'Accept': 'application/json,*/*', 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.87 Safari/537.36', 'Connection': 'keep-alive', 'Referer': 'http://localhost:8889/topologies/local/default/nlu-ex', 'Cookie': '_ga=GA1.1.59782163.1463778064'})
Traceback (most recent call last):
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/web.py", line 1334, in _execute
    result = yield result
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 631, in run
    yielded = self.gen.throw(*sys.exc_info())
  File "/Users/nlu/.pex/code/5674180e6b6e436e31a29124f90f1b52e728dcfa/heron/ui/src/python/handlers/api/metrics.py", line 85, in get
    results = yield futures
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 464, in callback
    result_list = [i.result() for i in children]
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 631, in run
    yielded = self.gen.throw(*sys.exc_info())
  File "/Users/nlu/.pex/code/5674180e6b6e436e31a29124f90f1b52e728dcfa/heron/ui/src/python/handlers/access/heron.py", line 566, in fetch
    results = yield futures
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 464, in callback
    result_list = [i.result() for i in children]
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 631, in run
    yielded = self.gen.throw(*sys.exc_info())
  File "/Users/nlu/.pex/code/5674180e6b6e436e31a29124f90f1b52e728dcfa/heron/ui/src/python/handlers/access/heron.py", line 366, in get_metrics
    raise tornado.gen.Return((yield fetch_url_as_json(request_url)))
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 631, in run
    yielded = self.gen.throw(*sys.exc_info())
  File "/Users/nlu/.pex/code/5674180e6b6e436e31a29124f90f1b52e728dcfa/heron/ui/src/python/handlers/access/fetch.py", line 45, in fetch_url_as_json
    http_response = yield tornado.httpclient.AsyncHTTPClient().fetch(fetch_url)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 111, in result
    raise self._exception
HTTPError: HTTP 404: Not Found
```

alanngai commented:
@nlu90, can you give me the steps to reproduce, as well as instructions on how to set things up so that I can see the page you' referring to?  The error is not obvious, so It's difficult for me to fix without having a working reference

nlu90 commented:
@alanngai sure. You'll need to build and start the heron-tracker/heron-ui on your localhost.

Step 1: build heron-tracker/heron-ui, in top dir of the heron project:
```
bazel run --config=darwin -- scripts/packages:heron-tools-install.sh --user
```

Step 2: start heron-tracker/heron-ui, run the following cmds in two separate terminals:
`heron-tracker` and `heron-ui`

Step 3: open a broswer and enter `localhost:8889`

Step 4: launch a local topology and check the output of heron-tracker/heron-ui terminal

kramasamy commented:
@alanngai - the documentation can be found in the web as well.

http://twitter.github.io/heron/docs/getting-started/

alanngai commented:
@kramasamy looks like the documentation on (http://twitter.github.io/heron/docs/developers/compiling/mac/) is dated.  bazel build is broken, and from the git log it looks like it's been upgraded to 0.2.3.  The documentation still has instructions to install 0.1.2

alanngai commented:
@nlu90, @kramasamy: i was able to build and run heron-tracker and heron-ui, but I was not able to launch a topology.  In fact, building heron per instructions (bazel build --config=darwin heron/...) fails for me with the following error

ERROR: /Users/alan/work/cloudmon/dev-master/heron-1/config/BUILD:5:1: Executing genrule //config:config-srcs failed: bash failed: error executing command 
  (cd /private/var/tmp/_bazel_alan/db3f53e50a1fc88e2445c3b62d2e0efd/heron-1 && \
  exec env - \
    PATH='/Applications/Xcode.app/Contents/Developer/usr/bin:~/dev/bin:/Users/alan/dev/activator:/Users/alan/dev/play:/Users/alan/dev/scala/bin:/Users/alan/dev/sbt/bin:/usr/local/mysql/bin:/Users/alan/dev/go/bin:/Users/alan/dev/apache-maven/bin:/Users/alan/dev/spark/bin:/Users/alan/dev/flink/bin:/Users/alan/dev/apache-ant/bin:/Users/alan/work/cloudmon/dev-master/Tools/DevTools:/Users/alan/dev/ec2-api-tools/bin:/Users/alan/bin:/Users/alan/.rvm/gems/ruby-1.9.3-p194/bin:/Users/alan/.rvm/gems/ruby-1.9.3-p194@global/bin:/Users/alan/.rvm/rubies/ruby-1.9.3-p194/bin:/Users/alan/.rvm/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Users/alan/.rvm/bin' \
    TMPDIR=/var/folders/fq/fzcyqkcx7rgbycg4kr8z3m180000gn/T/ \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; export WORKSPACE_ROOT=$(pwd)
export INSTALL_DIR=$(pwd)/bazel-out/local-fastbuild/genfiles/config
export TMP_DIR=$(mktemp -d -t config.XXXXX)
mkdir -p $TMP_DIR
cp -R config/m4/acx_pthread.m4 config/m4/gxx_stdio_filebuf.m4 config/autogen.sh config/configure.ac config/heron.def $TMP_DIR
cd $WORKSPACE_ROOT
cd $TMP_DIR
touch Makefile.am
mkdir -p m4
mv *.m4 m4
./autogen.sh
./configure --enable-shared=no
cp heron-config.h $INSTALL_DIR
rm -rf $TMP_DIR'): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 2.
Preparing the heron build system...please wait

Found GNU Autoconf version 2.69
Found GNU Automake version 1.15
Found GNU Libtool version 2.4.4

Warning: Several files expected of projects that conform to the GNU
coding standards were not found.  The files were automatically added
for you since you do not have a 'foreign' declaration specified.

Considered adding 'foreign' to AM_INIT_AUTOMAKE in configure.ac
or to AUTOMAKE_OPTIONS in your top-level Makefile.am file.


Automatically preparing build ... Warning: autoreconf failed
Attempting to run the preparation steps individually

Preparing build ... ERROR: aclocal failed
Target //scripts/packages:binpkgs failed to build
ERROR: /Users/alan/work/cloudmon/dev-master/heron-1/heron/common/src/cpp/errors/BUILD:32:1 Executing genrule //config:config-srcs failed: bash failed: error executing command 
  (cd /private/var/tmp/_bazel_alan/db3f53e50a1fc88e2445c3b62d2e0efd/heron-1 && \
  exec env - \
    PATH='/Applications/Xcode.app/Contents/Developer/usr/bin:~/dev/bin:/Users/alan/dev/activator:/Users/alan/dev/play:/Users/alan/dev/scala/bin:/Users/alan/dev/sbt/bin:/usr/local/mysql/bin:/Users/alan/dev/go/bin:/Users/alan/dev/apache-maven/bin:/Users/alan/dev/spark/bin:/Users/alan/dev/flink/bin:/Users/alan/dev/apache-ant/bin:/Users/alan/work/cloudmon/dev-master/Tools/DevTools:/Users/alan/dev/ec2-api-tools/bin:/Users/alan/bin:/Users/alan/.rvm/gems/ruby-1.9.3-p194/bin:/Users/alan/.rvm/gems/ruby-1.9.3-p194@global/bin:/Users/alan/.rvm/rubies/ruby-1.9.3-p194/bin:/Users/alan/.rvm/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Users/alan/.rvm/bin' \
    TMPDIR=/var/folders/fq/fzcyqkcx7rgbycg4kr8z3m180000gn/T/ \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; export WORKSPACE_ROOT=$(pwd)
export INSTALL_DIR=$(pwd)/bazel-out/local-fastbuild/genfiles/config
export TMP_DIR=$(mktemp -d -t config.XXXXX)
mkdir -p $TMP_DIR
cp -R config/m4/acx_pthread.m4 config/m4/gxx_stdio_filebuf.m4 config/autogen.sh config/configure.ac config/heron.def $TMP_DIR
cd $WORKSPACE_ROOT
cd $TMP_DIR
touch Makefile.am
mkdir -p m4
mv *.m4 m4
./autogen.sh
./configure --enable-shared=no
cp heron-config.h $INSTALL_DIR
rm -rf $TMP_DIR'): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 2.

nlu90 commented:
@alanngai did you merge with the most recent master?

alanngai commented:
yes, the error I posted was after the merge.  I merged again just now, same problem.  I even cloned a fresh copy of the project just to be sure, but ran into the same problem

alan@Leung-Sheung-Wing-Chun ~/work/cloudmon/dev-master/heron (master) $ bazel build --config=darwin heron/...
Warning: ignoring _JAVA_OPTIONS in environment.
.
SLF4J: The following set of substitute loggers may have been accessed
SLF4J: during the initialization phase. Logging calls during this
SLF4J: phase were not honored. However, subsequent logging calls to these
SLF4J: loggers will work as normally expected.
SLF4J: See also http://www.slf4j.org/codes.html#substituteLogger
SLF4J: org.eclipse.aether.internal.impl.DefaultRepositorySystem
INFO: Found 333 targets...
ERROR: /Users/alan/work/cloudmon/dev-master/heron/config/BUILD:5:1: Executing genrule //config:config-srcs failed: bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 2.
Preparing the heron build system...please wait

Found GNU Autoconf version 2.69
Found GNU Automake version 1.15
Found GNU Libtool version 2.4.4

Warning: Several files expected of projects that conform to the GNU
coding standards were not found.  The files were automatically added
for you since you do not have a 'foreign' declaration specified.

Considered adding 'foreign' to AM_INIT_AUTOMAKE in configure.ac
or to AUTOMAKE_OPTIONS in your top-level Makefile.am file.


Automatically preparing build ... Warning: autoreconf failed
Attempting to run the preparation steps individually

Preparing build ... ERROR: aclocal failed
INFO: Elapsed time: 13.354s, Critical Path: 6.44s

nlu90 commented:
@alanngai 

1. Can you try to install the 0.2.3 bazel again? You can down it from [here](https://github.com/bazelbuild/bazel/releases/tag/0.2.3). And pick either `bazel-0.2.3-installer-darwin-x86_64.sh` or `bazel-0.2.3-installer-linux-x86_64.sh` to install based on your machine type.

2. Before build the heron repo, did you run `./bazel_configure.py` to setup the environment correctly?

That's the two things came into my mind.

kramasamy commented:
@alanngai - can you do the following?

```
bazel build --config=darwin --verbose_failures heron/...
```

This will prevent verbose messages when a failure occurs in bazel.

nlu90 commented:
@alanngai Is there any update? I checked out the branch alan/pylint-rcfile from your heron fork(the latest commit is: `commit 2376b6dea135e9296e8b4163c7c7ed642181f72e`), and built with bazel 0.2.3 on my own mac. It worked fine:

```
[tw-mbp-nlu heron-1 (alan/pylint-rcfile)]$ bazel build --config=darwin heron/...
INFO: Found 331 targets...
...
...
...
Starting audit...
Audit done.
INFO: Elapsed time: 969.019s, Critical Path: 862.65s
```

alanngai commented:
reinstalled 0.2.3 via the link you gave me.  Still no luck.  Here's the output of ./bazel_configure.py

alan@Leung-Sheung-Wing-Chun ~/work/cloudmon/dev-master/heron (master)↑ $ ./bazel_configure.py 
Platform Darwin
Using C compiler          :	/Applications/Xcode.app/Contents/Developer/usr/bin/gcc (4.2.1)
Using C++ compiler        :	/Applications/Xcode.app/Contents/Developer/usr/bin/gcc (4.2.1)
Using C preprocessor      :	/usr/bin/cpp (7.3.0)
Using C++ preprocessor    :	/usr/bin/cpp (7.3.0)
Using linker              :	/Applications/Xcode.app/Contents/Developer/usr/bin/ld
Using Automake            :	/usr/local/Cellar/automake/1.15/bin/automake (1.15)
Using Autoconf            :	/usr/local/Cellar/autoconf/2.69/bin/autoconf (2.69)
Using Make                :	/Applications/Xcode.app/Contents/Developer/usr/bin/make (3.81)
Using CMake               :	/usr/local/Cellar/cmake/3.1.0/bin/cmake (3.1.0)
Using Python2             :	/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7 (2.7.10)
Using archiver            :	/usr/bin/libtool
Using coverage tool       :	/usr/bin/gcov
dwp                       :	not found, but ok
Using nm                  :	/usr/bin/nm
objcopy                   :	not found, but ok
objdump                   :	not found, but ok
Using strip               :	/usr/bin/strip
Wrote the environment exec file scripts/compile/env_exec.sh

here's the output of 'bazel build --config=darwin --verbose_failures heron/...'

 ./install-sh -c -d '/private/var/tmp/_bazel_alan/848753e5ba8b8f098ab31862e547561a/heron/bazel-out/local-fastbuild/genfiles/third_party/jansson/lib/pkgconfig'
 /usr/bin/install -c -m 644 jansson.pc '/private/var/tmp/_bazel_alan/848753e5ba8b8f098ab31862e547561a/heron/bazel-out/local-fastbuild/genfiles/third_party/jansson/lib/pkgconfig'
ERROR: /Users/alan/work/cloudmon/dev-master/heron/config/BUILD:5:1: Executing genrule //config:config-srcs failed: bash failed: error executing command 
  (cd /private/var/tmp/_bazel_alan/848753e5ba8b8f098ab31862e547561a/heron && \
  exec env - \
    PATH='/Applications/Xcode.app/Contents/Developer/usr/bin:~/dev/bin:/Users/alan/dev/activator:/Users/alan/dev/play:/Users/alan/dev/scala/bin:/Users/alan/dev/sbt/bin:/usr/local/mysql/bin:/Users/alan/dev/go/bin:/Users/alan/dev/apache-maven/bin:/Users/alan/dev/spark/bin:/Users/alan/dev/flink/bin:/Users/alan/dev/apache-ant/bin:/Users/alan/work/cloudmon/dev-master/Tools/DevTools:/Users/alan/dev/ec2-api-tools/bin:/Users/alan/bin:/Users/alan/.rvm/gems/ruby-1.9.3-p194/bin:/Users/alan/.rvm/gems/ruby-1.9.3-p194@global/bin:/Users/alan/.rvm/rubies/ruby-1.9.3-p194/bin:/Users/alan/.rvm/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Users/alan/.rvm/bin' \
    TMPDIR=/var/folders/fq/fzcyqkcx7rgbycg4kr8z3m180000gn/T/ \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; export WORKSPACE_ROOT=$(pwd)
export INSTALL_DIR=$(pwd)/bazel-out/local-fastbuild/genfiles/config
export TMP_DIR=$(mktemp -d -t config.XXXXX)
mkdir -p $TMP_DIR
cp -R config/m4/acx_pthread.m4 config/m4/gxx_stdio_filebuf.m4 config/autogen.sh config/configure.ac config/heron.def $TMP_DIR
cd $WORKSPACE_ROOT
cd $TMP_DIR
touch Makefile.am
mkdir -p m4
mv *.m4 m4
./autogen.sh
./configure --enable-shared=no
cp heron-config.h $INSTALL_DIR
rm -rf $TMP_DIR'): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 2.
Preparing the heron build system...please wait

Found GNU Autoconf version 2.69
Found GNU Automake version 1.15
Found GNU Libtool version 2.4.4

Warning: Several files expected of projects that conform to the GNU
coding standards were not found.  The files were automatically added
for you since you do not have a 'foreign' declaration specified.

Considered adding 'foreign' to AM_INIT_AUTOMAKE in configure.ac
or to AUTOMAKE_OPTIONS in your top-level Makefile.am file.


Automatically preparing build ... Warning: autoreconf failed
Attempting to run the preparation steps individually

Preparing build ... ERROR: aclocal failed
INFO: Elapsed time: 72.917s, Critical Path: 66.74s

alanngai commented:
I'm running OSX El Capitan 10.11.5

kramasamy commented:
@alanngai - I checkout a fresh version of your branch in my iMac running El Capitan 10.11.3 and it was successful

```
NFO: From Linking third_party/yaml-cpp/libyaml-cxx.a:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: bazel-out/local-fastbuild/bin/third_party/yaml-cpp/_objs/yaml-cxx/third_party/yaml-cpp/empty.pic.o has no symbols
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/local-fastbuild/bin/third_party/yaml-cpp/libyaml-cxx.a the table of contents is empty (no object file members in the library define global symbols)
INFO: From Executing extra_action //tools/java:checkstyle_java on //heron/schedulers/tests/java:HeronMasterDriverTest:
Starting audit...
Audit done.
INFO: From Executing extra_action //tools/java:checkstyle_java on //heron/examples/src/java:examples-unshaded:
Starting audit...
Audit done.
INFO: Elapsed time: 498.984s, Critical Path: 489.77s
```

kramasamy commented:
@alanngai - perhaps you should checkout a new version of your branch and try it out.

alanngai commented:
@kramasamy that output was from a freshly cloned master.  It's probably some env diff on my machine.  Can you copy and paste the output of your bazel_configure.py?  Maybe there's a difference there that's breaking things on my machine

kramasamy commented:
Here you go 

```
Karthiks-iMac:heron karthikz$ ./bazel_configure.py 
Platform Darwin
Using C compiler          :	/usr/bin/gcc (4.2.1)
Using C++ compiler        :	/usr/bin/g++ (4.2.1)
Using C preprocessor      :	/usr/bin/cpp (7.3.0)
Using C++ preprocessor    :	/usr/bin/cpp (7.3.0)
Using linker              :	/usr/bin/ld
Using Automake            :	/opt/twitter/Cellar/automake/1.14.1/bin/automake (1.14.1)
Using Autoconf            :	/opt/twitter/Cellar/autoconf/2.69/bin/autoconf (2.69)
Using Make                :	/usr/bin/make (3.81)
Using CMake               :	/opt/twitter/Cellar/cmake/3.0.2/bin/cmake (3.0.2)
Using Python2             :	/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7 (2.7.10)
Using archiver            :	/usr/bin/libtool
Using coverage tool       :	/usr/bin/gcov
dwp                       :	not found, but ok
Using nm                  :	/usr/bin/nm
objcopy                   :	not found, but ok
objdump                   :	not found, but ok
Using strip               :	/usr/bin/strip
Wrote the environment exec file scripts/compile/env_exec.sh
```

alanngai commented:
hmm, that didn't help.  All my versions are at least as the same or newer than yours.  I guess I'm going to have to try this on a different laptop or debug the makefiles.  I'll have to do it later in the week after I'm back from traveling

alanngai commented:
@kramasamy or @nlu90, would you mind checking to see if you get this warning message in your build log somewhere?  For me it happens right before the failure

Warning: Several files expected of projects that conform to the GNU
coding standards were not found.  The files were automatically added
for you since you do not have a 'foreign' declaration specified.

Considered adding 'foreign' to AM_INIT_AUTOMAKE in configure.ac
or to AUTOMAKE_OPTIONS in your top-level Makefile.am file.


Automatically preparing build ... Warning: autoreconf failed

alanngai commented:
better yet, if you could email me your entire build log to alan@opsclarity.com

nlu90 commented:
@alanngai My macos version is: OS X Yosemite, 10.10.5. 

I'll send the whole build log to you via email. But here are the output of bazel and ./bazel_configuration.py cmd
```
[tw-mbp-nlu heron_main (master)]$ bazel version
.
Build label: 0.2.3
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue May 17 14:22:21 2016 (1463494941)
Build timestamp: 1463494941
Build timestamp as int: 1463494941
[tw-mbp-nlu heron_main (master)]$ ./bazel_configure.py 
Platform Darwin
Using C compiler          :	/usr/bin/clang (7.0.2)
Using C++ compiler        :	/usr/bin/clang++ (7.0.2)
Using C preprocessor      :	/usr/bin/cpp (7.0.2)
Using C++ preprocessor    :	/usr/bin/cpp (7.0.2)
Using linker              :	/usr/bin/ld
Using Automake            :	/opt/twitter/Cellar/automake/1.14.1/bin/automake (1.14.1)
Using Autoconf            :	/opt/twitter/Cellar/autoconf/2.69/bin/autoconf (2.69)
Using Make                :	/usr/bin/make (3.81)
Using CMake               :	/opt/twitter/Cellar/cmake/3.0.1/bin/cmake (3.0.1)
Using Python2             :	/opt/twitter/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/bin/python2.7 (2.7.8)
Using archiver            :	/usr/bin/libtool
Using coverage tool       :	/usr/bin/gcov
dwp                       :	not found, but ok
Using nm                  :	/usr/bin/nm
objcopy                   :	not found, but ok
objdump                   :	not found, but ok
Using strip               :	/usr/bin/strip
Wrote the environment exec file scripts/compile/env_exec.sh
```

alanngai commented:
interesting.  @nlu90 it looks like my bazel conf is different from yours.  You're using clang whereas I'm using gcc.  But @kramasamy is using gcc and that did not appear to be a problem for him.  The version numbers for all of my stuff is at least as up to date as yours.  I did find one difference in your build log that's different from mine.  Here's yours

-- @nlu90 ---------------------------------------
Found GNU Autoconf version 2.69
Found GNU Automake version 1.14.1

Warning:  libtoolize does not appear to be available.  This means that
the automatic build preparation via autoreconf will probably not work.
Preparing the build by running each step individually, however, should
work and will be done automatically for you if autoreconf fails.

Fortunately, glibtoolize was found which means that your system may simply
have a non-standard or incomplete GNU Autotools install.  If you have
sufficient system access, it may be possible to quell this warning by
running:

   sudo ln -s /opt/twitter/bin/glibtoolize /opt/twitter/bin/libtoolize

Found GNU Libtool version 2.4.2

Warning: Several files expected of projects that conform to the GNU
coding standards were not found.  The files were automatically added
for you since you do not have a 'foreign' declaration specified.

Considered adding 'foreign' to AM_INIT_AUTOMAKE in configure.ac
or to AUTOMAKE_OPTIONS in your top-level Makefile.am file.


Automatically preparing build ... done

-- @alanngai ---------------------------------------
Found GNU Autoconf version 2.69
Found GNU Automake version 1.15
Found GNU Libtool version 2.4.4

Warning: Several files expected of projects that conform to the GNU
coding standards were not found.  The files were automatically added
for you since you do not have a 'foreign' declaration specified.

Considered adding 'foreign' to AM_INIT_AUTOMAKE in configure.ac
or to AUTOMAKE_OPTIONS in your top-level Makefile.am file.


Automatically preparing build ... Warning: autoreconf failed

----------------------------------------------------

It could be that my new versions of automake or libtool is breaking things.  I suspect it might be automake.  Brew doesn't seem to allow me to revert to a prev version however.   This should probably be fixed, because if automake versioning is the problem, I probably won't be the only one with this problem.  Will need to investigate further when I get back to the office later this week.

In the meantime, @kramasamy, would you mind sending me your build log as well, so I can cross reference with @nlu90 and mine?  Thanks

alanngai commented:
@kramasamy, would you mind emailing me your build log?  Thanks

alanngai commented:
Tried to build on a fresh laptop with el capitan, got exactly the same result.  At this point, not sure what to do.  @nlu90, @kramasamy, if you guys get me additional build logs, that might help

/bin/sh ./libtool --tag=CXX   --mode=link g++ -D_THREAD_SAFE  -DGTEST_HAS_PTHREAD=1 -g -O2   -o lib/libgmock_main.la -rpath /private/var/tmp/_bazel_admin/47be9456eb3d2e2be5bf2a38cf8dcd18/heron/bazel-out/local-fastbuild/genfiles/third_party/gmock/lib src/gmock_main.lo lib/libgmock.la 
libtool: link: ar cru lib/.libs/libgmock_main.a  src/gmock_main.o
libtool: link: ranlib lib/.libs/libgmock_main.a
libtool: link: ( cd "lib/.libs" && rm -f "libgmock_main.la" && ln -s "../libgmock_main.la" "libgmock_main.la" )
ERROR: /Users/admin/work/heron/config/BUILD:5:1: Executing genrule //config:config-srcs failed: bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
Preparing the heron build system...please wait

Found GNU Autoconf version 2.69
Found GNU Automake version 1.15

Warning:  libtoolize does not appear to be available.  This means that
the automatic build preparation via autoreconf will probably not work.
Preparing the build by running each step individually, however, should
work and will be done automatically for you if autoreconf fails.

ERROR: Unable to locate GNU Libtool.

ERROR:  To prepare the heron build system from scratch,
        at least version 1.4.2 of GNU Libtool must be installed.

autogen.sh does not need to be run on the same machine that will
run configure or make.  Either the GNU Autotools will need to be installed
or upgraded on this system, or autogen.sh must be run on the source
code on another system and then transferred to here. -- Cheers!

INFO: Elapsed time: 30.728s, Critical Path: 24.78s

alanngai commented:
@kramasamy, @nlu90: never mind.  It looks like i was missing lib tool on the new laptop and the build proceeded successfully.  I should have a path forward

kramasamy commented:
@alanngai - great to hear. It will be great if you could merge the conflicts and update the PR. @nlu90 - can you another look at it so that we can merge this?

nlu90 commented:
@kramasamy once this PR is updated, I'll look into it

alanngai commented:
merged latest from upstream.  Once I got the build to work, I didn't have any issues with the UI.  Topo submission, activation, deactivation, and kill seemed to work fine

<img width="1052" alt="screen shot 2016-07-06 at 4 07 45 pm" src="https://cloud.githubusercontent.com/assets/1654394/16637506/58753088-4394-11e6-8379-8af3e11c2437.png">
<img width="1026" alt="screen shot 2016-07-06 at 4 07 07 pm" src="https://cloud.githubusercontent.com/assets/1654394/16637507/58764982-4394-11e6-89f9-32e3042f4877.png">

nlu90 commented:
Hi, @alanngai Based on the log output of heron-ui and heron-tracker. I found two problems:

The first one is from heon-ui:
```
Traceback (most recent call last):
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/web.py", line 1334, in _execute
    result = yield result
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 631, in run
    yielded = self.gen.throw(*sys.exc_info())
  File "/Users/nlu/.pex/code/b661095b1351239862044a4429b3d51c6d46d19d/heron/ui/src/python/handlers/api/metrics.py", line 52, in get
    results = yield futures
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 628, in run
    value = future.result()
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 464, in callback
    result_list = [i.result() for i in children]
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/concurrent.py", line 109, in result
    raise_exc_info(self._exc_info)
  File "/Users/nlu/.pex/install/tornado-4.0.2-py2.7-macosx-10.10-intel.egg.eb452d50fb45f95017438ce5b0db2edc504b3a97/tornado-4.0.2-py2.7-macosx-10.10-intel.egg/tornado/gen.py", line 175, in wrapper
    yielded = next(result)
  File "heron/common/src/python/handler/access/heron.py", line 396, in get_comp_metrics
    create_url(METRICS_URL_FMT),
TypeError: url_concat() takes exactly 2 arguments (1 given)
```

It's caused by a recent change from us, I'll fix it.

nlu90 commented:
The second one is from heron-tracker:
```
06 Jul 2016 18:40:10-WARNING:web.py:1811: 404 GET /topologies/metricsquery?cluster=local&environ=default&starttime=1467844800&query=RATE%28TS%28word%2C%2A%2C__jvm-gc-collection-time-ms%29%29&endtime=1467855600&topology=nlu-ex (::1) 95.23ms
Traceback (most recent call last):
.
.
.
  File "heron/tracker/src/python/metricstimeline.py", line 114, in getMetricsTimeline
    raise Exception(metricResponse.status.message)
Exception: Unknown component
```

@cckellogg Can you help have a look?

objmagic commented:
looks good to me. I see that config file specifies two-space indent, which is good. Can you make sure that all the other configs also conform to Twitter's Python coding style [here](https://github.com/OpsClarity/heron-1/blob/f0faa74415193ad575504d0077c97d740279709a/tools/python/checkstyle.ini)?

alanngai commented:
@objmagic, which configs are you referring to?  Also, if it's an addition, can we merge this pr first and handle the additional changes in a separate ticket?  I'll be happy to do it if you point me to which configs you mean.

objmagic commented:
@alanngai sorry for the confusion. I mean [checkstyle.ini](https://github.com/OpsClarity/heron-1/blob/f0faa74415193ad575504d0077c97d740279709a/tools/python/checkstyle.ini).

alanngai commented:
@objmagic, changed checkstyle.ini indent to 2 spaces

objmagic commented:
@alanngai there is some confusion going on. I mean it is good that you followed Twitter's Python coding rule and enforced indentation level to 2 at here: https://github.com/OpsClarity/heron-1/blob/078dbad4f627079f8369749990c58937cddbddeb/tools/python/checkstyle.ini#L266 . I was not talking about `checkstyle.ini` itself.

sorry about the confusion...

alanngai commented:
@objmagic, so you're not asking for additional changes?

objmagic commented:
@alanngai no. just want to make sure [your config file](https://github.com/OpsClarity/heron-1/blob/f0faa74415193ad575504d0077c97d740279709a/tools/python/checkstyle.ini) has correct config options, according to [Twitter's Python style](https://github.com/OpsClarity/heron-1/blob/f0faa74415193ad575504d0077c97d740279709a/tools/python/checkstyle.ini)

alanngai commented:
@objmagic, I see, but the two links in your comment point to the same file?  

Also, if there are to be additional checkstyle changes, I'd recommend that we handle it in a separate ticket and just merge this PR in first.  I've already had to fix the same styling issues on the same files several times because of having to merge from upstream over the course of a month and a half.  It was @kramasamy's recommendation when I started work on this to just go with the default pylint settings with  some indent rule tweaks, as a first pass

objmagic commented:
@alanngai sorry. updated.

alanngai commented:
created #1053 to track @objmagic's suggesting

kramasamy commented:
Alan - we are working through the last bug. If that clears, we will merge this PR in. 

> On Jul 8, 2016, at 5:06 PM, Alan Ngai <notifications@github.com> wrote:
> 
> created #1053 to track @objmagic's suggesting
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cckellogg commented on heron/common/src/python/handler/access/heron.py:
```diff
@@ -11,6 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
```
I reverted this file and got it to work. For this patch can we revert this file and then try to improve it in another patch?

objmagic replied:
why reverting this file can make style checker work?

objmagic commented:
@alanngai 

Our goal after merging this PR is to enable Python style checker. Checkout #587 and you can see currently [we disable Python style checker in CI](https://github.com/twitter/heron/pull/587/files#diff-4c3a4e61471476647ad1c0987a744414R39).

You can refer to #587 and run 

`bazel build --config=darwin --experimental_action_listener=tools/python:compile_python heron/cli/src/python/...`. 

We want this command have exit status 0.

kramasamy commented:
@alanngai - just to clarify what @objmagic and @cckellogg pointed out

- Merge with the current master

- There is an exception that occurs in tracker which @cckellogg pointed out to reverting the changes in a single file. Let us revert this first or fix the error occurring in this file.

- @objmagic mentioned that there are still python style checker errors occurring in your code. Could you run the style checker again and please fix those?

- Once these are done, we will merge the code and then address the changes to enable python checker by default.

alanngai commented:
@kramasamy 

1. I'll take a look at fixing or reverting heron.py later today 
1. wrt python checkstyle errors, there are 2 classes of errors
  - explicitly ignored checkstyle errors
  - errors that I don't enough context to fix (labelled with FIXME), one of which I've filed an issue for (#855)

in other words, i can tackle (1) but not (2)

kramasamy commented:
@alanngai - thanks let us do 1) first and merge it. I will request @objmagic to take care of 2).

objmagic commented:
@kramasamy @alanngai @cckellogg @nlu90 

As for 1), we are actually talking about two errors:
1. https://github.com/twitter/heron/pull/849#issuecomment-230960262, which is fixed by @nlu90 
2. https://github.com/twitter/heron/pull/849#issuecomment-230960639, which we had no idea in this morning. 

However, I just realized I have seen the same error message before. I checked with @nlu90 this afternoon and we believe that this error happens because metrics info is refreshed every one minute. That being said, if you fire up ``heron-tracker`` and ``heron-ui`` immediately after you submit the topology, this exception will show up. **If you refresh the UI after one minute you open the `heron-ui`, you will not see any error message.**

Another reason that makes me more confident in my conclusion is that @alanngai 's change on `heron.py` is purely syntax level change and involves no change on the semantic of the program. If we really want to consider the above scenario in the above paragraph an issue, we can merge this issue first, and I can submit a new PR to make sure no exception will be thrown even if there is no metrics information.

objmagic commented:
@kramasamy and yes, I can look into 2) after we merge this.

kramasamy commented:
@objmagic - regarding the exception trace for 2) - can you fix in a separate PR and send a pull request?

objmagic commented:
@alanngai @kramasamy 

I think we can go ahead and merge this PR now.

I will keep track of related issues after merging this PR. Namely:
1. fix https://github.com/twitter/heron/pull/849#issuecomment-231932982
2. fix https://github.com/twitter/heron/pull/849#issuecomment-231834711 ("errors that I don't have enough context to fix (labelled with FIXME)")
3. #578 #579 #580 #581 #582 #583 #584

nlu90 commented:
👍 

@objmagic @kramasamy let's merge this PR and fix the problem in a new PR
